### PR TITLE
[Performance][Model] DeepSeek-V4: gather logits rows before hc_head on prefill steps

### PR DESCRIPTION
**Pull request by TeloySXH**
## Summary
Head-path latency at T=8192/L=1: **0.417 ms → 0.049 ms (8.47×, kernel-level)**;
E2E TTFT unchanged within noise (head path ≈ 0.04% of a prefill step).

## Why

On DeepSeek-V4, the final head runs `hc_head_fused_kernel_tilelang` + a final
RMSNorm on **all T rows** of the `[T, hc, H]` hidden states, and the runner
gathers the logits rows afterwards via `hidden_states[logits_indices]`. On
DSpark prefill-containing steps only a few rows per request are actually
sampled, so every other row is computed only to be discarded (e.g. T=512
with 1-8 logits rows).

## What

Gather `hidden_states[logits_indices]` to `[L, hc, H]` **before** hc_head +
RMSNorm when the batch only needs the logits rows. Both ops are row-wise, so
the selected rows match the full-width path up to rare 1-bf16-ULP RMSNorm
reduction-order differences across row counts.

Gated by `VLLM_MOE_HEAD_GATHER_FIRST` (default on; env off restores the old
path) and restricted to CPU-known-safe eager steps:
- `cg_mode == CUDAGraphMode.NONE` (captured graphs bake shapes/addresses and
  replay without `model_inputs`);
- prefill-containing batches (`has_prefill`) on the last PP rank;
- no pooling, no PCP manager, no prompt-logprobs requests, no batch-sharded
  sampling;
- speculators other than dspark/dflash excluded (eagle3 unverified).

## Tests

- `tests/models/test_deepseek_v4_head_gather.py` (7 cases): hc_head rows are
  **bitwise identical** gathered-before vs gathered-after; the full
  hc_head+RMSNorm chain stays within the pinned envelope
  (`MAX_MISMATCH_RATE = 1e-4`) with a per-element 1-ULP check.
- Kernel-level microbench (T=8192, L=1): head-tail latency 0.417 ms → 0.049 ms
  (**8.47×**); production cold prefill tail: hc_head 358 µs → ~7 µs.
- 20/20 multi-turn smoke; TTFT within 1% across three runs.

> Baseline note: benchmark/E2E numbers were measured on the internal
> vLLM-MoE fork (v0.28 line, 4×H20/sm_90) where this change has been in
> production use; the unit-test file is ported as-is. Compile-level and gate
> logic checks were run on this branch; GPU revalidation on v0.29 main
> is pending.

## Duplicate check

GitHub PR search (`head_gather`, `hc_head`, DSv4+logits) returns no existing
PR addressing this change.

---
AI assistance was used in preparing this change (see Co-authored-by trailer
in the commit).

### CODE REVIEW

**Comment by TeloySXH** [comment]
Hi maintainers 👋 — some context on this change:

This optimization was developed and validated on our internal vLLM-MoE fork running DeepSeek-V4-Flash-0731 on a single node with 4× H20 (sm_90, v0.28 line) where it has been in production use and works well. On a frozen benchmark (think=high), the head-path kernel latency at T=8192/L=1 drops from 0.417 ms → 0.049 ms (8.47×) and the production cold-prefill tail (hc_head) goes from ~358 µs to ~7 µs; E2E TTFT stays within ~1% across three runs, which is expected since the head path accounts for only ~0.04% of a prefill step. The 7-case unit test (bitwise hc_head equivalence + pinned 1-ULP envelope) passed, and a 20-turn smoke test completed 20/20.

Since this PR is ported from the v0.28 line to current main, I would very much welcome a reproduction of these measurements on upstream test hardware, as a way to independently confirm both the effectiveness and the correctness of the change on the current codebase.

This is my first contribution to vLLM, and I would love to get more involved in the project open-source community going forward. I am happy to make any revisions the reviewers suggest — please let me know what you think.😋😋🥰

**Comment by TeloySXH** [comment]
> **More detail on the testing** (done on the internal vLLM-MoE fork, v0.28 line; this branch is the port to current main):
>
> **Environment**
> - Hardware: single node, 4× NVIDIA H20 (sm_90a, 96 GiB/card, NVLink P2P), TP4 + Expert Parallel
> - Model: DeepSeek-V4-Flash-0731 (~284B MoE, 156 GB checkpoint), DSpark speculative decoding K=7, probabilistic draft, adaptive verification ON
> - Engine: vLLM 0.28.1.dev4 (editable from the internal fork), torch 2.13.0+cu129, CUDA 12.8 toolkit; KV cache FP8, block_size 256, max_model_len 32768, gpu_memory_utilization 0.90, CUDA Graph FULL_AND_PIECEWISE (max capture 512), breakable PIECEWISE enabled
>
> **Test plan (5 stages)**
> - **M0 Correctness**: new 7-case unit test (hc_head gathered-before vs gathered-after is **bitwise identical**; full hc_head+RMSNorm chain within a pinned ≤1-ULP envelope, mismatch rate ≤1e-4); greedy-output archives + acceptance-rate interval comparison (three-arm)
> - **M1 Kernel microbench** (`bench_hc_head_tail.py`, warmup 10 / iter 50): tail latency before/after across the T×L matrix
> - **M2 torch profiler, cold prefill** (unique prompt to bypass prefix-cache; chunked into 8186+2603): hc_head/norm/gather timing and its share η of the step
> - **M3 E2E three-arm A/B**: baseline / same code with `VLLM_MOE_HEAD_GATHER_FIRST=0` / default-on; 4 scenarios (S1_P512, S1_P8192, C8_P512, C8_P8192), 3 runs each, median
> - **M4 Stability**: graph capture, warmup failures=0, 36 bench runs all rc=0, 20-turn multi-turn smoke 20/20 on both arms
>
> **What the process caught**
> - A first naive profile sampled the **APC-hot chunk** (shared prompt + `--seed-once` cached 10496/10767 tokens); re-measuring with a unique prompt gave true cold-prefill numbers, which then matched the microbench exactly.
> - Our deployment auto-enables **breakable PIECEWISE** graphs for T≤512 prefill/mixed steps; captured graphs replay without `model_inputs`, so the gate requires `cg_mode == NONE` (found during implementation review).
> - The chain is not bitwise-stable: `vllm_c` RMSNorm chooses different reduction orders for different row counts (3/262144 elements, 1 bf16 ULP); `hc_head` itself is bitwise identical per row — hence the pinned 1-ULP envelope.
> - Greedy text diff is not usable as a gate here: the same code run twice differs in 6/9 cases (inherent MoE/attention reduction-order nondeterminism), with divergences in the decode mid-section where this change is inactive.
>
> **Results (M1)**
>
> | T | L | before | after | speedup |
> | ---: | ---: | ---: | ---: | ---: |
> | 8192 | 1 | 0.417 ms | 0.049 ms | **8.47×** |
> | 8192 | 8 | 0.413 ms | 0.050 ms | 8.33× |
> | 8192 | 64 | 0.413 ms | 0.050 ms | 8.27× |
> | 2048 | 1 | 0.118 ms | 0.050 ms | 2.38× |
> | 512 | 1 | 0.049 ms | 0.049 ms | ≈1× (launch-bound) |
>
> **M2**: cold prefill (T=8186 chunk): hc_head **358.0 → 7.7 µs**, tail (head+norm+gather) **412.7 → 7.7 µs**, hc_head grid `[8186] → [1]`. η ≈ **0.04%** of the step.
> **M3**: E2E TTFT within **1%** across all four scenarios (no regression); decode tok/s & acceptance drift between arms is at the same ~10% level as baseline-vs-`off` environment noise.
> **Memory**: intermediate materialization drops from 2×[T,4096] bf16 (**128 MiB** at T=8192) to [L,…] (48 KiB at L=1).
>
> We'd be glad to share the benchmark script, and a reproduction on upstream test hardware/current main would be very welcome to independently confirm both the effectiveness and the correctness.

**Comment by TeloySXH** [comment]
Thanks for the thorough review — all three points are addressed in the latest commit fce8a2a:
MTP slice (CodeRabbit): the drafter buffer is now sliced by input_batch.num_tokens_after_padding instead of hidden_states.shape[0], so with gather-first the drafter still receives the full-width T pre-hc_head rows (identical to pre-change behavior; with the flag off the two values are the same).
Model scoping (Copilot): the gate now requires DeepseekV4ForCausalLM, so other models never receive the logits_indices kwarg; microbatched forward (ubatch_runner) is also excluded.
Test skip (Copilot): the unit test now skips when TileLang is unavailable, not only on non-CUDA platforms.
🥰🥰

**Comment by TeloySXH** [comment]
## Local revalidation on current main (this branch)

**PR:** #55573 (`fix/dsv4-head-logits-gather-first` @ `fce8a2a`)  
**Goal:** GPU revalidation after the v0.28 → current-main port (as noted in the PR body).  
**Not a duplicate of the internal MoE-fork M1–M4 numbers.**

### Environment

| item | value |
| --- | --- |
| Host | 4× NVIDIA H20 (sm_90, 96 GiB), driver 535.161.08 + cuda-compat-12-9 |
| Python | 3.12.14 isolated venv (`/dockerdata/vllm-t-env`), not conda base |
| torch | 2.13.0+cu129 |
| vLLM | editable this tree (`0.1.dev20916+gdc02934a0.precompiled`) |
| Model (smoke) | DeepSeek-V4-Flash-0731, TP4+EP, DSpark K=7 |

Caches/JIT stayed on local XFS. No changes on `main`.

### 1. GPU — gather-first numerical contract

```bash
source env_isolated.sh
python -m pytest --noconftest -v --tb=short \
  tests/models/test_deepseek_v4_head_gather.py
```

**Result: 7 passed** (5.84s).

| case | result |
| --- | --- |
| `test_hc_head_gather_first_bitwise` T=512/2048 × L=1/64 | 4 passed, bitwise |
| `test_head_norm_chain_all_rows_bitwise` L=T=256 | passed, bitwise |
| `test_head_norm_chain_within_one_ulp` T=2048/8192, L=64 | 2 passed (≤1-ULP envelope) |

`--noconftest` because this tree's `tests/conftest.py` deps are not fully installed.

### 2. Serve smoke (eager + DSpark, so gather-first can fire)

```bash
# isolated env, port 8004 only; adaptive verification off (incompatible with cudagraph NONE)
vllm serve /dockerdata/model/DeepSeek-v4-Flash-0731 \
  --host 127.0.0.1 --port 8004 \
  --tensor-parallel-size 4 --enable-expert-parallel \
  --kv-cache-dtype fp8 --block-size 256 --max-model-len 4096 \
  --gpu-memory-utilization 0.90 \
  --tokenizer-mode deepseek_v4 --trust-remote-code \
  --reasoning-parser deepseek_v4 \
  --speculative-config '{"method":"dspark","num_speculative_tokens":7,"draft_sample_method":"probabilistic","attention_backend":"FLASH_ATTN","enable_adaptive_verification":false}' \
  --compilation-config '{"cudagraph_mode":"NONE"}' \
  --kernel-config '{"enable_jit_warmup":false}' \
  --disable-custom-all-reduce
```

`cudagraph_mode=NONE` is required for this check: gather-first is gated to eager prefill steps (`cg_mode == NONE` + `has_prefill`). Startup ~3 min on this host.

One greedy request after `Application startup complete`:

```bash
curl -sS http://127.0.0.1:8004/v1/chat/completions \
  -H 'Content-Type: application/json' \
  -d '{"model":"/dockerdata/model/DeepSeek-v4-Flash-0731","messages":[{"role":"user","content":"Reply with the single word: ok"}],"max_tokens":16,"temperature":0}'
```

**Result: HTTP 200.** `choices[0].message.content == "ok"`, `finish_reason=stop`, usage `prompt=90 completion=15` (12 reasoning tokens). No `ERROR` / `Traceback` in the serve log.

Stopped with `./close.sh` (port 8004 only).

Fork M1–M4 (kernel microbench / cold-prefill profiler / three-arm E2E / 20-turn) were measured on the **v0.28 fork** (Adaptive ON, FULL_AND_PIECEWISE). This tree is correctness + smoke only; I did not rerun that bench.

Dummy-run MTP still slices `get_mtp_target_hidden_states()` by `hidden_states.shape[0]`; the real `propose()` path already uses `input_batch.num_tokens_after_padding`. Not treating that as a reval blocker.

### Ask

Please add the `ready` label (or `/ci run`) when convenient. I cannot add it (0 merged PRs).

**Comment by TeloySXH** [comment]
Hi @zyongye  @WoosukKwon  👋

This is my first contribution to vLLM, and I would love to keep contributing to the project going forward. I'd really appreciate your review of this one when you have a moment.

**What it does:** On DeepSeek-V4 prefill steps, the final head runs `hc_head_fused_kernel_tilelang` + RMSNorm over all `T` rows even though only a few logits rows per request are sampled. This PR gathers `hidden_states[logits_indices]` first, so those ops run on `[L, hc, H]` instead. Gated by `VLLM_MOE_HEAD_GATHER_FIRST` (default on, env off restores the old path) and limited to CPU-known-safe eager steps, so no other model or spec/ubatch path is affected.

**Results:** head-path kernel 0.417 ms → 0.049 ms (8.47×) at T=8192/L=1; E2E TTFT within noise (the head path is ~0.04% of a prefill step). I also revalidated on current main with 4×H20 and posted the details in the thread.

I addressed the bot review feedback: MTP hidden-state slice width (CodeRabbit), model-scoping the gate to `DeepseekV4ForCausalLM` and adding a TileLang-MHC skip guard to the tests (Copilot). Happy to iterate on anything you'd prefer changed — e.g. splitting it further, adjusting the gate, or adding more tests.

One practical note: since this is a fork PR and I don't yet have merged PRs in the repo yet, `pre-run-check` is failing and CI is blocked. Could you add the `ready` label when you get a chance so the checks can run? Thank you for your time! 🙏

**Comment by TeloySXH** [comment]
Rebased onto latest `main` (`07b755346`) to clear the merge conflict / `needs-rebase` label.
The only textual conflict was in `vllm/v1/worker/gpu/input_batch.py`: both sides added a field after `max_query_len`. The rebase keeps both — `head_gather_first` from this PR and `fast_prefill` from #56145. The gather-first gate now also requires `fast_prefill is None`, so the KV-sharing fast-prefill path is not stacked with head gather. `model_runner.py` / `model.py` / `envs.py` auto-merged.
GPU recheck on 4×H20 after the rebase:

**Comment by TeloySXH** [comment]
Rebased onto current main (2c2cbd84). The only textual conflict was in gpu/model_runner.py: main added watermarking prepare_watermarking next to the MTP hidden-state slice. The rebase keeps both — gather-first still slices the pre-hc_head buffer by num_tokens_after_padding when the head has already been reduced to L logits rows, and the watermarking hook is unchanged. Gate, tests, and model path were otherwise a clean replay of the previous review-feedback commit.
